examples/asyncio/router_router.py

```diff
@@ -0,0 +1,75 @@
+"""
+ROUTER-ROUTER communication example
+
+ideal for P2P applications that need a socket that is able to connect
+to several other peers simultaneously, while also being able
+to receive messages
+
+Contributed by github:jcpinto54 and github:frobnitzem
+"""
+# Copyright (c) 2022 PyZMQ Developers.
+# This example is in the public domain (CC-0)
+
+import asyncio
+
+from aiowire import EventLoop
+
+import zmq
+from zmq.asyncio import Context
+
+
+class Server:
+    def __init__(self, url):
+        context = Context.instance()
+        socket = context.socket(zmq.ROUTER)
+        socket.setsockopt_string(zmq.IDENTITY, 'server')
+        socket.bind(url)
+        self.socket = socket
+
+    async def run(self, ev):
+        # `Wire` type receives the event loop as its single argument
+
+        req = await self.socket.recv_multipart()
+        print(f"Server received {req}")
+        # identity of receptionist, empty frame, message content
+        await self.socket.send_multipart([req[0], b'', b'whatup'])
+
+        # continue to run by returning this `Wire`
+        return self.run
+
+
+class Client:
+    def __init__(self, url, name):
+        self.name = name
+
+        context = Context.instance()
+        socket = context.socket(zmq.ROUTER)
+        socket.setsockopt_string(zmq.IDENTITY, name)
+        socket.connect(url)
+        self.socket = socket
+
+    async def run(self, ev):
+        await asyncio.sleep(0.1)
+        # identity of receptionist, empty frame, message content
+        await self.socket.send_multipart([b'server', b'', b'cheers'])
+        rep = await self.socket.recv_multipart()
+        print(f"{self.name} received {rep}")
+
+        # continue to run by returning this `Wire`
+        return self.run
+
+
+url = 'inproc://test_zmq'
+srv = Server(url)
```

Comment:
Given the pending deprecation of globals like `asyncio.get_event_loop()`, I'd like to encourage creating asyncio objects within the `async def main`

Reply:
I don't _think_ that any of the parts of the `Client` or `Server` `__init__` functions call `get_event_loop` - so Client and Server classes are not themselves asyncio objects - unless zmq.asyncio.Context is an asyncio object.

I guess the worry is that some code outside of an async function might call Socket.recv or Server.run, which would be incorrect.

Reply:
That's true for pyzmq >= 22.2, but pyzmq 22.1 called `get_event_loop` in . I think it's still a pattern to encourage: do not create module-global async objects, even if creating the objects themselves _may_ not yet grab a handle on the loop.